### PR TITLE
Add symbolic manipulation of BQM objects

### PR DESCRIPTION
**Pull request by arcondello**
Add very simple operator overloading to the BQM object.

#### Alternatives Considered
Right now `__add__` etc create a new BQM object, rather than something like
```
Add = namedtuple('Add', ['lhs', 'rhs'])

class BQM:
    ...
    def __add__(self, other):
        return Add(self, other)
```

Returning a relation object like above has performance advantages (at the cost of memory), but it's a lot less flexible and a lot less self-contained. For instance, right now
```
ss = ExactSolver().sample(bqm0 + bqm1)
```
is defined without needing some sort of resolution step.

One way around this would be to make comparisons full-fledged BQM data classes (e.g. [pyBQM](https://github.com/dwavesystems/dimod/blob/be747fa4be2c898f7460f77573f82d35f634df57/dimod/binary/pybqm.py#L35) or [VartypeView](https://github.com/dwavesystems/dimod/blob/be747fa4be2c898f7460f77573f82d35f634df57/dimod/binary/vartypeview.py#L56)) and then do something like
```
class BQM:
    ...
    def __add__(self, other):
        bqm = self.__new__(type(self))
        bqm.data = AddBQM(self, other)
```

but for now I think it makes sense to keep things simple.

#### Additional Context

Performance of the multiplication operator would be improved by #833 
Closes #790 as redundant.

### CODE REVIEW

File: dimod/binary/binary_quadratic_model.py [hunk 1]
```diff
@@ -390,30 +392,161 @@ def __deepcopy__(self, memo):
...
+    def __add__(self, other: Union['BinaryQuadraticModel', Bias]):
+        # in python 3.8+ we could do this is functools.singledispatchmethod
+        if isinstance(other, BinaryQuadraticModel):
+            if other.num_variables and other.vartype != self.vartype:
+                # future: return QuadraticModel
+                raise TypeError("cannot add BQMs with different vartypes")
+            new = self.copy()
+            new.update(other)
+            return new
+        if isinstance(other, Number):
```
Comment (randomir) [inline dimod/binary/binary_quadratic_model.py:414]: Doesn't `DictBQM` support `object` for bias `dtype`?

Currently `NotImplemented` is raised, even if that object implements `+`.

In other places as well.

File: dimod/sym.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        except AttributeError:
+            return False
+
+
+class Ge(Comparison, sense='>='):
+    pass
+
+
+class Le(Comparison, sense='<='):
+    pass
```
Comment (randomir) [inline dimod/sym.py:60]: You could compare lhs.offset with numeric rhs. Probably doesn't make much sense, but it seems consistent.

**Review by randomir — APPROVED** [review]
LGTM!

File: dimod/sym.py (hunk 2 above)
Comment (arcondello) [reply dimod/sym.py:60]: IMO, makes sense to keep the implementation simpler. Though if we get a feature request for that I am happy to support it.

File: dimod/binary/binary_quadratic_model.py (hunk 1 above)
Comment (arcondello) [reply dimod/binary/binary_quadratic_model.py:414]: Hmm, you're right. And I just checked that `isinstance(sympy.Symbol('a'), Number) == False`... I'll take a look.

Comment (arcondello) [reply dimod/binary/binary_quadratic_model.py:414]: Ok, I played with a couple different variations of trying to handle this case and this way lies madness. I think (for now) only allowing symbolic manipulation for numeric types strikes a good balance between nice sugar and not having to deal with a million edge cases.

Users can always make use of the `bqm.offset += val` and `bqm.scale(val)` directly, which will handle other types reasonably well (at least when combined with #624 )

Comment (randomir) [reply dimod/binary/binary_quadratic_model.py:414]: Yeah, among other things, a bias can be just another BQM :laughing:

**Pull request merged**